File: spec/javascript/publisher_spec.js
var jstest = require("jstest").Test

var Publisher = require("../../src/mixins/publisher"),
    assign    = require("../../src/util/assign")

jstest.describe("Publisher", function() { with(this) {
  before(function() { with(this) {
    this.publisher = assign({}, Publisher)
  }})

  describe("with subscribers that remove themselves", function() { with(this) {
    before(function() { with(this) {
      this.calledA = false
      this.calledB = false

      this.handler = function() {
        calledA = true
        publisher.unbind("event", handler)
      }

      publisher.bind("event", handler)
      publisher.bind("event", function() { calledB = true })
    }})

    it("successfully calls all the callbacks", function() { with(this) {
      publisher.trigger("event")
      assert( calledA )
      assert( calledB )
    }})
  }})
}})

File: spec/ruby/publisher_spec.rb
require "spec_helper"

describe Faye::Publisher do
  let(:publisher) { Class.new { include Faye::Publisher }.new }

  describe "with subscribers that remove themselves" do
    before do
      @called_a = false
      @called_b = false

      handler = lambda do
        @called_a = true
        publisher.unbind(:event, &handler)
      end

      publisher.bind(:event, &handler)
      publisher.bind(:event) { @called_b = true }
    end

    it "successfully calls all the callbacks" do
      publisher.trigger(:event)
      @called_a.should == true
      @called_b.should == true
    end
  end
end
